server/src/main/scala/com/cloudera/livy/server/LivyServer.scala

```diff
@@ -135,6 +167,54 @@ class LivyServer extends Logging {
     sys.props("livy.server.serverUrl") = _serverUrl.get
   }
 
+  def runKinit(keytab: String, principal: String): Boolean = {
+    val commands = Seq("kinit", "-kt", keytab, principal)
+    val proc = new ProcessBuilder(commands: _*).inheritIO().start()
+    proc.waitFor() match {
+      case 0 =>
+        debug("Ran kinit command successfully.")
+        kinitFailCount = 0
+        UserGroupInformation.getCurrentUser.reloginFromTicketCache()
+        true
+      case _ =>
+        warn("Fail to run kinit command.")
+        kinitFailCount += 1
+        false
+    }
+  }
+
+  def startKinitDaemon(keytab: String, principal: String): Unit = {
+    executor.schedule(
+      new Runnable() {
+        def run(): Unit = {
+          if (runKinit(keytab, principal)) {
+            // schedule another kinit run with a fixed delay.
+            executor.schedule(
+              new Runnable() {
+                def run(): Unit = {
+                  runKinit(keytab, principal)
```

Comment:
So, if I read things correctly, this will login once, then login a second time after one hour, and then stop.

Don't you need to schedule `this` instead of this different runnable?